### ISSUE TITLE
Scanning action isn't cleaned up correctly on loss of Bluetooth connectivity, resulting in crash

### ISSUE THREAD

**Issue by akmarinov** [steamclock/bluejay#187]
Summary:

I get the following crash

![screenshot 2019-01-06 17 17 05](https://user-images.githubusercontent.com/1358638/50737924-edb13d00-11d6-11e9-8d4a-b34fd07ab497.png)

Which seems to result in the file Queue.swift on line 130
```
        if !queue.isEmpty {
            preconditionFailure("Queue is not emptied at the end of cancel all.")
        }
```

The queue at the time contains a scan action, which wasn't cleaned up properly after this code:

```
         for queueable in queue where !queueable.state.isFinished {
            queueable.fail(error)

            if let connection = queueable as? Connection {
                if !connection.state.isFinished {
                    log("Interrupting cancel all to terminate a pending connection...")
                    return
                }
            }
        }
```

It seems that when this doesn't cancel the Scan action and the `stopScanning` method is called, the logic to stop and unqueue the scan action goes to the Scan.swift's `stopScan` method, which calls upon the `updateQueue` method, which contains the following:

```
func update(cancel: Bool = false, cancelError: Error? = nil) {
        if queue.isEmpty {
            log("Queue is empty, nothing to update.")
            return
        }

        if !isCBCentralManagerReady {
            log("Queue is paused because CBCentralManager is not ready yet.")
            return
        }

        ...
```

Thus when you transition from an active state to a no bluetooth state, this update method returns early, since it checks the Bluetooth connectivity and neglects to continue on and cancel the scan action.

This causes the queue to wind up with at least one scan action, failing the precondition in the Queue.swift file.

Steps to Reproduce:

1. Turn on bluetooth an initiate a scan
2. Turn off bluetooth from the device's settings, while a scan is queued
 
or

1. On a simulator initiate a scan without checking for bluetooth connectivity (the lack of bluetooth immediately triggers the issue)

Actual Result:

Both result in the precondition on Queue.swift:130 being triggered

Expected Result: 

The framework should clear the scan action and the queue on either the `updateQueue` method being called or the `cancellAll` method, which should clear up the crash.

Device, Build, OS:
iPhone X iOS 12.1, Bluejay v.0.7.1
Xcode 10.1 (10B61), Bluejay v.0.7.1, iPhone X simulator iOS 12.1 (16B91)

**Comment by akmarinov** [comment]
I see now that it'll potentially be addressed in https://github.com/steamclock/bluejay/pull/183

**Comment by JeremyChiang** [comment]
This is now fixed in v0.8.0. You can test, debug, and verify using the demo app in Bluejay.

When Bluetooth is toggled off, the logs show the scan being stopped with the correct error, removed from the queue, and released from memory.

![img_1386](https://user-images.githubusercontent.com/4675020/51143879-7d24a480-1804-11e9-8d53-e9d54148bc5e.png)

Then after toggling Bluetooth back on, the scan is resumed correctly. Both cases indicate the queue is either emptied or empty when it should be empty, therefore not hitting that precondition unexpectedly anymore.

![img_1387](https://user-images.githubusercontent.com/4675020/51143888-831a8580-1804-11e9-8567-cf84fc451e5b.png)
